File: app/src/main/java/org/ole/planet/myplanet/ui/team/teamDiscussion/DiscussionListFragment.java
package org.ole.planet.myplanet.ui.team.teamDiscussion;

import android.content.res.Configuration;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputLayout;
import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.model.RealmNews;
import org.ole.planet.myplanet.model.RealmTeamNotification;
import org.ole.planet.myplanet.ui.news.AdapterNews;
import org.ole.planet.myplanet.ui.team.BaseTeamFragment;
import org.ole.planet.myplanet.utilities.Constants;
import org.ole.planet.myplanet.utilities.FileUtils;
import org.ole.planet.myplanet.utilities.Utilities;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

import io.realm.Sort;

public class DiscussionListFragment extends BaseTeamFragment {
    RecyclerView rvDiscussion;
    TextView tvNodata;

    public DiscussionListFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_discussion_list, container, false);
        v.findViewById(R.id.add_message).setOnClickListener(view -> showAddMessage());
        rvDiscussion = v.findViewById(R.id.rv_discussion);
        tvNodata = v.findViewById(R.id.tv_nodata);
        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        List<RealmNews> realmNewsList = getNews();
        int count = realmNewsList.size();
        mRealm.executeTransactionAsync(realm -> {
            RealmTeamNotification notification = realm.where(RealmTeamNotification.class).equalTo("type", "chat").equalTo("parentId", teamId).findFirst();
            if (notification == null) {
                notification = realm.createObject(RealmTeamNotification.class, UUID.randomUUID().toString());
                notification.setParentId(teamId);
                notification.setType("chat");
            }
            notification.setLastCount(count);
        });
        changeLayoutManager(getResources().getConfiguration().orientation, rvDiscussion);
        showRecyclerView(realmNewsList);
    }

    private List<RealmNews> getNews() {
        List<RealmNews> realmNewsList = mRealm.where(RealmNews.class).isEmpty("replyTo").sort("time", Sort.DESCENDING).findAll();
        List<RealmNews> list = new ArrayList<>();
        for (RealmNews news : realmNewsList) {
            if (!TextUtils.isEmpty(news.getViewableBy()) && news.getViewableBy().equalsIgnoreCase("teams") && news.getViewableId().equalsIgnoreCase(team.get_id())) {
                list.add(news);
            } else if (!TextUtils.isEmpty(news.getViewIn())) {
                JsonArray ar = new Gson().fromJson(news.getViewIn(), JsonArray.class);
                for (JsonElement e : ar) {
                    JsonObject ob = e.getAsJsonObject();
                    if (ob.get("_id").getAsString().equalsIgnoreCase(team.get_id())) {
                        list.add(news);
                    }
                }
            }
        }
        return list;
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        changeLayoutManager(newConfig.orientation, rvDiscussion);
    }

    private void showRecyclerView(List<RealmNews> realmNewsList) {
        AdapterNews adapterNews = new AdapterNews(getActivity(), realmNewsList, user, null);
        adapterNews.setmRealm(mRealm);
        adapterNews.setListener(this);
        rvDiscussion.setAdapter(adapterNews);
        showNoData(tvNodata, adapterNews.getItemCount());
    }

    private void showAddMessage() {
        View v = getLayoutInflater().inflate(R.layout.alert_input, null);
        TextInputLayout layout = v.findViewById(R.id.tl_input);
        v.findViewById(R.id.add_news_image).setOnClickListener(vi -> FileUtils.openOleFolder(this, 100));
        v.findViewById(R.id.ll_image).setVisibility(Constants.showBetaFeature(Constants.KEY_NEWSADDIMAGE, getActivity()) ? View.VISIBLE : View.GONE);
        layout.setHint(getString(R.string.enter_message));
        new AlertDialog.Builder(getActivity()).setView(v).setTitle(getString(R.string.add_message)).setPositiveButton(getString(R.string.save), (dialogInterface, i) -> {
            String msg = layout.getEditText().getText().toString().trim();
            if (msg.isEmpty()) {
                Utilities.toast(getActivity(), getString(R.string.message_is_required));
                return;
            }
            HashMap<String, String> map = new HashMap<>();
            map.put("viewInId", teamId);
            map.put("viewInSection", "teams");
            map.put("message", msg);
            map.put("messageType", team.getTeamType());
            map.put("messagePlanetCode", team.getTeamPlanetCode());
            RealmNews.createNews(map, mRealm, user, imageList);
            Utilities.log("discussion created");
            rvDiscussion.getAdapter().notifyDataSetChanged();
        }).setNegativeButton(getString(R.string.cancel), null).show();
    }

    @Override
    public void setData(List<RealmNews> list) {
        showRecyclerView(list);
    }

}

File: app/src/main/java/org/ole/planet/myplanet/ui/team/teamDiscussion/DiscussionListFragment.kt
package org.ole.planet.myplanet.ui.team.teamDiscussion

import android.content.DialogInterface
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import com.google.gson.Gson
import com.google.gson.JsonArray
import io.realm.Realm
import io.realm.RealmResults
import io.realm.Sort
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.AlertInputBinding
import org.ole.planet.myplanet.databinding.FragmentDiscussionListBinding
import org.ole.planet.myplanet.model.RealmNews
import org.ole.planet.myplanet.model.RealmNews.Companion.createNews
import org.ole.planet.myplanet.model.RealmTeamNotification
import org.ole.planet.myplanet.ui.news.AdapterNews
import org.ole.planet.myplanet.ui.team.BaseTeamFragment
import org.ole.planet.myplanet.utilities.Constants
import org.ole.planet.myplanet.utilities.Constants.showBetaFeature
import org.ole.planet.myplanet.utilities.FileUtils.openOleFolder
import org.ole.planet.myplanet.utilities.Utilities
import java.util.UUID

class DiscussionListFragment : BaseTeamFragment() {
    private lateinit var fragmentDiscussionListBinding: FragmentDiscussionListBinding
    private var updatedNewsList: RealmResults<RealmNews>? = null
    private var filteredNewsList: List<RealmNews?> = listOf()
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        fragmentDiscussionListBinding = FragmentDiscussionListBinding.inflate(inflater, container, false)
        fragmentDiscussionListBinding.addMessage.setOnClickListener { showAddMessage() }

        updatedNewsList = mRealm.where(RealmNews::class.java).isEmpty("replyTo").sort("time", Sort.DESCENDING).findAllAsync()

        updatedNewsList?.addChangeListener { results ->
            filteredNewsList = filterNewsList(results)
            setData(filteredNewsList)
        }
        return fragmentDiscussionListBinding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val realmNewsList = news
        val count = realmNewsList.size
        mRealm.executeTransactionAsync { realm: Realm ->
            var notification = realm.where(RealmTeamNotification::class.java).equalTo("type", "chat").equalTo("parentId", teamId).findFirst()
            if (notification == null) {
                notification = realm.createObject(RealmTeamNotification::class.java, UUID.randomUUID().toString())
                notification.parentId = teamId
                notification.type = "chat"
            }
            notification?.lastCount = count
        }
        changeLayoutManager(resources.configuration.orientation, fragmentDiscussionListBinding.rvDiscussion)
        showRecyclerView(realmNewsList)
    }

    private fun filterNewsList(results: RealmResults<RealmNews>): List<RealmNews?> {
        val filteredList: MutableList<RealmNews?> = ArrayList()
        for (news in results) {
            if (!TextUtils.isEmpty(news.viewableBy) && news.viewableBy.equals("teams", ignoreCase = true) && news.viewableId.equals(team._id, ignoreCase = true)) {
                filteredList.add(news)
            } else if (!TextUtils.isEmpty(news.viewIn)) {
                val ar = Gson().fromJson(news.viewIn, JsonArray::class.java)
                for (e in ar) {
                    val ob = e.asJsonObject
                    if (ob["_id"].asString.equals(team._id, ignoreCase = true)) {
                        filteredList.add(news)
                    }
                }
            }
        }
        return filteredList
    }

    private val news: List<RealmNews>
        get() {
            val realmNewsList: List<RealmNews> = mRealm.where(RealmNews::class.java).isEmpty("replyTo").sort("time", Sort.DESCENDING).findAll()
            val list: MutableList<RealmNews> = ArrayList()
            for (news in realmNewsList) {
                if (!TextUtils.isEmpty(news.viewableBy) && news.viewableBy.equals("teams", ignoreCase = true) && news.viewableId.equals(team._id, ignoreCase = true)) {
                    list.add(news)
                } else if (!TextUtils.isEmpty(news.viewIn)) {
                    val ar = Gson().fromJson(news.viewIn, JsonArray::class.java)
                    for (e in ar) {
                        val ob = e.asJsonObject
                        if (ob["_id"].asString.equals(team._id, ignoreCase = true)) {
                            list.add(news)
                        }
                    }
                }
            }
            return list
        }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        changeLayoutManager(newConfig.orientation, fragmentDiscussionListBinding.rvDiscussion)
    }

    private fun showRecyclerView(realmNewsList: List<RealmNews?>?) {
        val adapterNews = activity?.let {
            realmNewsList?.let { it1 -> AdapterNews(it, it1.toMutableList(), user, null) }
        }
        adapterNews?.setmRealm(mRealm)
        adapterNews?.setListener(this)
        fragmentDiscussionListBinding.rvDiscussion.adapter = adapterNews
        if (adapterNews != null) {
            showNoData(fragmentDiscussionListBinding.tvNodata, adapterNews.itemCount, "discussions")
        }
        adapterNews?.notifyDataSetChanged()
    }

    private fun showAddMessage() {
        val binding = AlertInputBinding.inflate(layoutInflater)
        val layout = binding.tlInput
        binding.addNewsImage.setOnClickListener {
            val openFolderIntent: Intent = openOleFolder()
            openFolderLauncher.launch(openFolderIntent)
        }
        binding.llImage.visibility = if (showBetaFeature(Constants.KEY_NEWSADDIMAGE, requireContext())) View.VISIBLE else View.GONE
        layout.hint = getString(R.string.enter_message)
        AlertDialog.Builder(requireActivity())
            .setView(binding.root)
            .setTitle(getString(R.string.add_message))
            .setPositiveButton(getString(R.string.save)) { _: DialogInterface?, _: Int ->
                val msg = "${layout.editText?.text}".trim { it <= ' ' }
                if (msg.isEmpty()) {
                    Utilities.toast(activity, getString(R.string.message_is_required))
                    return@setPositiveButton
                }
                val map = HashMap<String?, String>()
                map["viewInId"] = teamId
                map["viewInSection"] = "teams"
                map["message"] = msg
                map["messageType"] = team.teamType ?: ""
                map["messagePlanetCode"] = team.teamPlanetCode ?: ""
                user?.let { createNews(map, mRealm, it, imageList) }
                fragmentDiscussionListBinding.rvDiscussion.adapter?.notifyDataSetChanged()
                setData(news)
            }
            .setNegativeButton(getString(R.string.cancel), null)
            .show()
    }

    override fun setData(list: List<RealmNews?>?) {
        showRecyclerView(list)
    }
}
